### ISSUE TITLE
Vault Distribution Edge Case

### ISSUE THREAD

**Issue by ashmbanks** [coordinape/coordinape#1836]
**Summary:**

1. **What is actually happening:**
- vault distribution Tx submitted (and saved to our DB)
- Tx failed, and was replaced with new Tx (with higher gas) by user
- Our system didn't know about the second Tx that suceeded
- **RESULT: Distribution happened (and funds are gone from vault), but app doesn't know about distribution or claims (and users can't claim funds on the app)** 
2. **What should be happening:**
App should be aware of the second Tx and process properly so distribution, claims, and therefore funds are all available.

Discord thread of incident investigation: https://discord.com/channels/845400066532704256/1065672789702475796

**Comment by crabsinger** [comment]
Example txs of this happening:

low gas tx that was never processed: https://etherscan.io/tx/0x87f4fd7d3ef30439b766c395c590acead5ed942b4730e9b7661432b0d17e1dfc

replaced by tx with higher gas: [0x11325ed38c96a54c64006878f4e1800a6212d5958bcf411349632ccf4339ede6](https://etherscan.io/tx/0x11325ed38c96a54c64006878f4e1800a6212d5958bcf411349632ccf4339ede6)

**Comment by topocount** [comment]
# Proposed Solution

1. don't drop distribution events when the tx is null.
2. receive a pending distribution event with a null tx object.
3. find the first block timestamp since the pending event was posted
4. query all blocks, including that one for the `EpochFunded` Event from our contract.
5. Compare the merkle roots. If one matches, follow the fulfillment of the distribution using the new tx hash, and delete the pending vault tx row.

**Comment by crabsinger** [comment]
Another Propsed Solution:

1. Query etherscan for the pending tx: https://etherscan.io/tx/0x87f4fd7d3ef30439b766c395c590acead5ed942b4730e9b7661432b0d17e1dfc
2. Parse the returned body for a `Replaced by Txn Hash` line, and parse out that tx hash.

3. Verify that tx hash matches our distribution with merkle root and other matching data.
4. If it all matches, use it as the new txHash
5. Donezo.

**Comment by levity** [comment]
In Ethers.js, transaction.wait is supposed to throw an error when the transaction is replaced -- see https://docs.ethers.org/v5/api/providers/types/#providers-TransactionResponse

So then we could catch this error and update the pending_transaction row with the new tx hash, and switch to calling wait() on the new tx

**Comment by topocount** [comment]
> In Ethers.js, transaction.wait is supposed to throw an error when the transaction is replaced -- see https://docs.ethers.org/v5/api/providers/types/#providers-TransactionResponse
> 
> So then we could catch this error and update the pending_transaction row with the new tx hash, and switch to calling wait() on the new tx

@levity The issue with this proposal is that our frontend doesn't detect when a tx is replaced, so we submit a tx hash to our backend that will probably get dropped (since it's fair to infer that the newer candidate probably has a higher gas price). By the time we process the recovery in our backend, the tx object cannot be retrieved _at all_ since dropped transactions return `null` when queried by [getTransaction](https://docs.ethers.org/v5/api/providers/provider/#Provider-getTransaction). This leaves us with no pointer (such as a nonce) to find the processed transaction and we have to scan blocks from the `created_at` time in the db.

**Comment by levity** [comment]
I meant that we would catch the replacement on the frontend.

But it’s true that we would have to augment the backend code too, for e.g.
the sequence “submit tx from frontend, close browser, replace tx from
wallet”. In that case, if ethers’ getTransaction doesn’t work with dropped
txs, then yeah, we’ll have to find something else. But I would be surprised
if we had to resort to scraping etherscan’s website instead of at least
using an api

**Comment by crabsinger** [comment]
If you can find an API, please let us know. I looked a bit and was unable
to find one.

**Comment by levity** [comment]
#1887 provides only a frontend solution

**Comment by levity** [comment]
Yes, it's in my backlog, but it's as you said, there's no obvious API
(Etherscan's API support is startlingly weak), and there are higher
priorities atm
